lib/active_model/serializer/adapter/json.rb

```diff
@@ -46,6 +22,111 @@ def serializable_hash(options = nil)
         def fragment_cache(cached_hash, non_cached_hash)
           Json::FragmentCache.new().fragment_cache(cached_hash, non_cached_hash)
         end
+
+        private
+
+        # iterate through the associations on the serializer,
+        # adding them to the parent as needed (as singular or plural)
+        #
+        # nested_associations is a list of symbols that governs what
+        # associations on the passed in seralizer to include
+        def add_resource_relationships(parent, serializer, nested_associations = [])
+          # have the include array normalized
+          nested_associations = ActiveModel::Serializer::Utils.include_array_to_hash(nested_associations)
+
+          included_associations = if nested_associations.present?
+            serializer.associations.select{ |association|
+              # nested_associations is a hash of:
+              #   key => nested association to include
+              nested_associations.has_key?(association.name)
+            }
+          else
+            serializer.associations
+          end
+
+          included_associations.each do |association|
+            serializer = association.serializer
+            opts = association.options
+            key = association.key
+
+            # sanity check if the association has nesting data
+            has_nesting = nested_associations[key].present?
+            if has_nesting
+              include_options_from_parent = { include: nested_associations[key] }
+              opts = opts.merge(include_options_from_parent)
+            end
+
+            if serializer.respond_to?(:each)
+              parent[key] = add_relationships(serializer, opts)
+            else
+              parent[key] = add_relationship(serializer, opts)
+            end
+          end
+
+          parent
+        end
+
+        # add a singular relationship
+        # the options should always belong to the serializer
+        def add_relationship(serializer, options)
```

Comment:
Can we maybe re-use some of the naming from https://github.com/rails-api/active_model_serializers/blob/1388ae82f26133d62f4e6cdfc40d1c659e121935/lib/active_model/serializer/adapter/json_api.rb#L91-L113 ?